File: modules/analysis/templates/03-regression/04-causal/causal_models.py
#!/usr/bin/env python3
"""Run configured causal identification strategies."""

from __future__ import annotations

import argparse
import sys
from pathlib import Path


def _load_common():
    for parent in Path(__file__).resolve().parents:
        if (parent / "_shared").exists():
            sys.path.insert(0, str(parent / "_shared"))
            import python_common  # type: ignore

            return python_common
    raise RuntimeError("Cannot locate templates/_shared/python_common.py")


common = _load_common()


def main() -> int:
    parser = common.add_common_args(argparse.ArgumentParser(description=__doc__))
    args = parser.parse_args()
    p = common.paths(args.out_root)
    data_path = Path(args.data) if args.data else p["data"] / "analysis-data.csv"
    if not data_path.exists():
        common.log_run(args, "03-regression/04-causal", "blocked", [], f"缺少数据：{data_path}")
        return 2
    df = common.load_data(data_path)
    roles = common.infer_roles(df, args.dict_path or p["data"] / "variable-dictionary.csv")
    models = {}
    blockers = []
    for name, fn in [("IV/2SLS", common.run_iv_2sls), ("DID/TWFE", common.run_twfe_did), ("RDD", common.run_parametric_rdd), ("IPW", common.run_ipw_ate)]:
        try:
            models[name] = fn(df, roles)
        except Exception as exc:
            blockers.append(f"{name}: {exc}")
    if not models:
        report = common.write_markdown(p["reports"] / "causal-blockers.md", "Causal Blockers", {"阻断": "\n".join(blockers)})
        common.log_run(args, "03-regression/04-causal", "blocked", [report], "没有可执行的识别策略。")
        return 2
    table = common.export_regression_table(models, p["tables"] / "tableA1-causal-robustness.csv", roles.controls, roles.fe, roles.cluster, "Table A1 Causal Robustness")
    summary = common.export_model_summary(p["reports"] / "causal-results.md", models, "Causal Results")
    balance = common.write_csv(p["tables"] / "matching-balance.csv", [{"method": "IPW/PSM/CEM", "status": "若 IPW 成功，权重回归已纳入 tableA1；PSM/CEM 需按项目依赖扩展。"}])
    common.log_run(args, "03-regression/04-causal", "ok", [table, summary, balance], "已运行可配置的 IV/DID/RDD/IPW。")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: modules/analysis/templates/03-regression/04-causal/causal_models.R
# Run configured causal identification strategies.

`%||%` <- function(a, b) if (is.null(a) || length(a) == 0) b else a
find_shared <- function() {
  cmd <- commandArgs(FALSE); file_arg <- sub("^--file=", "", cmd[grep("^--file=", cmd)][1] %||% "")
  p <- if (nzchar(file_arg)) dirname(normalizePath(file_arg, mustWork = FALSE)) else getwd()
  repeat { candidate <- file.path(p, "_shared", "r_common.R"); if (file.exists(candidate)) return(candidate); np <- dirname(p); if (identical(np, p)) stop("Cannot locate templates/_shared/r_common.R"); p <- np }
}
source(find_shared())
args <- parse_common_args(); ensure_dirs(args$out_root)
data_path <- if (!is.na(args$data) && file.exists(args$data)) args$data else file.path(args$out_root, "data", "analysis-data.csv")
if (!file.exists(data_path)) { log_run(args, "03-regression/04-causal", "blocked", "-", paste("缺少数据:", data_path)); quit(status = 2) }
df <- load_data(data_path); roles <- infer_roles(df, args$dict)
mods <- list(); blockers <- c()
mods$IV <- tryCatch(run_iv_2sls(df, roles), error = function(e) { blockers <<- c(blockers, paste("IV:", e$message)); NULL })
mods$DID <- tryCatch(run_twfe_did(df, roles), error = function(e) { blockers <<- c(blockers, paste("DID:", e$message)); NULL })
mods$RDD <- tryCatch(run_rdrobust(df, roles), error = function(e) { blockers <<- c(blockers, paste("RDD:", e$message)); NULL })
mods$IPW <- tryCatch(run_weightit_ipw(df, roles), error = function(e) { blockers <<- c(blockers, paste("IPW:", e$message)); NULL })
mods <- mods[!vapply(mods, is.null, logical(1))]
if (!length(mods)) {
  report <- write_md(file.path(args$out_root, "reports", "causal-blockers.md"), "Causal Blockers", c("阻断" = paste(blockers, collapse = "\n")))
  log_run(args, "03-regression/04-causal", "blocked", report, "没有可执行的识别策略。")
  quit(status = 2)
}
table <- export_models(mods, file.path(args$out_root, "tables", "tableA1-causal-robustness.csv"), "Table A1 Causal Robustness", roles$controls, roles$fe, roles$cluster)
summary_path <- write_md(file.path(args$out_root, "reports", "causal-results.md"), "Causal Results", lapply(mods, function(m) paste(capture.output(summary(m)), collapse = "\n")))
log_run(args, "03-regression/04-causal", "ok", c(table, summary_path), "已运行可配置的 IV/DID/RDD/IPW。")
